An older version of our analyses (used in the preprint's liger integrations) used the final tables `marm_human_orthologs.csv` and `marm_mouse_orthologs.csv` (included in this repo), generated as follows:
A single ortholog table downloaded from Biomart (https://ensembl.org/biomart/martview/) as of ENSEMBL v110 using marmoset ("White-tufted-ear marmoset genes") as the reference species and selecting `Attributes` > `Homologues` to find orthologs from mouse and human. An NCBI gene ID conversion table was also downloaded from Biomart. The two final ortholog tables were generated by running the scripts `make_marm_human_orthologs.py` and `make_marm_mouse_orthologs.py`, which use the two downloaded tables as inpiuts.

However, all the other (and current) analyses use a single ortholog table downloaded from ENSEMBL v112 using human (GRCh38.p14) as the initial reference organism (`ens_112_ortho_human_ferret_pig_mouse_marmoset.csv`). 
